test/unit/ui_utils_spec.js

```diff
@@ -398,4 +399,208 @@ describe('ui_utils', function() {
+      const views = [];
+      const { scrollLeft, scrollTop, } = scroll;
+      const scrollRight = scrollLeft + scroll.clientWidth;
+      const scrollBottom = scrollTop + scroll.clientHeight;
+      for (const view of pages) {
+        const { div, } = view;
+        const viewLeft = div.offsetLeft + div.clientLeft;
+        const viewRight = viewLeft + div.clientWidth;
+        const viewTop = div.offsetTop + div.clientTop;
+        const viewBottom = viewTop + div.clientHeight;
+
+        if (viewLeft < scrollRight && viewRight > scrollLeft &&
+            viewTop < scrollBottom && viewBottom > scrollTop) {
+          const hiddenHeight = Math.max(0, scrollTop - viewTop) +
+                               Math.max(0, viewBottom - scrollBottom);
+          const percent =
+            ((div.clientHeight - hiddenHeight) * 100 / div.clientHeight) | 0;
+          views.push({ id: view.id, x: viewLeft, y: viewTop, view, percent, });
+        }
+      }
+      return { first: views[0], last: views[views.length - 1], views, };
+    }
+
+    // This function takes a fixed layout of pages and compares the system under
+    // test to the slower implementation above, for a range of scroll viewport
+    // sizes and positions.
+    function scrollOverDocument(pages) {
+      const size = pages.reduce(function(max, { div, }) {
+        return Math.max(max, div.offsetTop + div.clientTop + div.clientHeight);
+      }, 0);
+      for (let i = 0; i < size; i += 7) {
+        for (let j = i + 5; j < size; j += (j - i)) {
+          const scroll = {
+            scrollTop: i,
+            scrollLeft: 0,
+            clientHeight: j - i,
+            clientWidth: 10000,
+          };
+          expect(getVisibleElements(scroll, pages))
+            .toEqual(slowGetVisibleElements(scroll, pages));
+        }
+      }
+    }
+
+    it('with pages of varying height', function() {
+      const pages = makePages([
+        [[50, 20], [20, 50]],
+        [[30, 12], [12, 30]],
+        [[20, 50], [50, 20]],
+        [[50, 20], [20, 50]],
+      ]);
+      scrollOverDocument(pages);
+    });
+    it('widescreen challenge', function() {
+      const pages = makePages([
+        [[10, 50], [10, 60], [10, 70], [10, 80], [10, 90]],
+        [[10, 90], [10, 80], [10, 70], [10, 60], [10, 50]],
+        [[10, 50], [10, 60], [10, 70], [10, 80], [10, 90]],
+      ]);
+      scrollOverDocument(pages);
+    });
+
+    // These tests refer to cases enumerated in the comments of
+    // backtrackBeforeAllVisibleElements.
+    it('handles backtrack case 1', function() {
+      const pages = makePages([
+        [[10, 20], [10, 20], [10, 20], [10, 20]],
+        [[10, 50], [10, 10], [10, 50], [10, 10]],
+        [[10, 50], [10, 50], [10, 50], [10, 50]],
+        [[10, 20], [10, 20], [10, 20], [10, 20]],
+        [[10, 20]],
+      ]);
+      const scroll = {
+        scrollTop: 70,
+        scrollLeft: 0,
+        clientHeight: 30,
```

Comment:
De-magicking these numbers is a difficult thing to do while maintaining clarity. Is this the sort of thing you want?
```js
const scroll = {
  scrollTop:
    20 + SPACING + // height of the first row
    40, // a value between 30 (so the 10-height pages on the second row
        // are hidden) and 50 (so the 50-height pages are visible)
  scrollLeft: 0,
  clientHeight:
    10 + // the deficit left by the 40 above
    SPACING + // completing the second row
    10, // any value greater than zero, so that the third row pages are
        // visible
  clientWidth: 10000,
};
```

Or, going ‘full constant’, I could write something like this (which to me is pretty horrifying):

```js
const FIRST_ROW_HEIGHT = 20;
const SECOND_ROW_HEIGHT = 50;
const SECOND_ROW_SHORTER_HEIGHT = 10;
const THIRD_ROW_HEIGHT = 50;
const MORE_PAGES_HEIGHT = 20;
const pages = makePages([
  [
    [10, FIRST_ROW_HEIGHT],
    [10, FIRST_ROW_HEIGHT],
    [10, FIRST_ROW_HEIGHT],
    [10, FIRST_ROW_HEIGHT],
  ],
  [
    [10, SECOND_ROW_HEIGHT],
    [10, SECOND_ROW_SHORTER_HEIGHT],
    [10, SECOND_ROW_HEIGHT],
    [10, SECOND_ROW_SHORTER_HEIGHT],
  ],
  [
    [10, THIRD_ROW_HEIGHT],
    [10, THIRD_ROW_HEIGHT],
    [10, THIRD_ROW_HEIGHT],
    [10, THIRD_ROW_HEIGHT],
  ],
  [
    [10, MORE_PAGES_HEIGHT],
    [10, MORE_PAGES_HEIGHT],
    [10, MORE_PAGES_HEIGHT],
    [10, MORE_PAGES_HEIGHT],
  ],
  [
    [10, MORE_PAGES_HEIGHT],
  ],
]);

// must be between 0 and SECOND_ROW_HEIGHT / 2 -
//   SECOND_ROW_SHORTER_HEIGHT / 2, which happens to be 20
const SMALL_NUMBER = 10;

// an arbitary positive number
const ANOTHER_SMALL_NUMBER = 15;

const scroll = {
  scrollTop:
    FIRST_ROW_HEIGHT + SPACING +
    SECOND_ROW_HEIGHT / 2 + SECOND_ROW_SHORTER_HEIGHT / 2 + SMALL_NUMBER,
  scrollLeft: 0,
  clientHeight:
    SECOND_ROW_HEIGHT / 2 - SECOND_ROW_SHORTER_HEIGHT / 2 - SMALL_NUMBER +
    SPACING + ANOTHER_SMALL_NUMBER;
  clientWidth: 10000,
};
```

Reply:
Perhaps a better thing to do than changing these numbers would be to positively assert in the test that pages 4, 6, and 8 through 11 are visible (zero-indexed); that would make it clear what case we're expecting to be testing, even if the numbers that generate that case remain mysterious.

Reply:
Definitely not the bottom one. It is too verbose and adds little clarity.

The first example looks much better. When I wrote the comment, I was thinking of a variable to show which pages are important, something like this (edit comments or variable names if you deem necessary):

```
      // A page in the second row whose height is taller than the others.
      // The other (smaller) pages in the row are vertically centered,
      // with gaps evenly distributed above and below them.
      const BIG_PAGE = [10, 50];
      const pages = makePages([
        [[10, 20], [10, 20], [10, 20], [10, 20]],
        [[10, 50], [10, 10], BIG_PAGE, [10, 10]], // scroll's top intersects here
        [[10, 50], [10, 50], [10, 50], [10, 50]], // scroll's bottom intersects here
        [[10, 20], [10, 20], [10, 20], [10, 20]],
        [[10, 20]],
      ]);
```

The `BIG_PAGE` variable is just an example, you could also put `50` in a separate variable or something. (I choose `BIG_PAGE` because that is somewhat descriptive and still maintains the nice alignment of the matrix).

and then some commentary or calculation that shows how the numbers are determined (basically what you did in the first example).

In my review, I will be looking for answers to:

- When I look at the test, is it immediately obvious what the test is supposedly doing?
   (For the current test, the answer is no, because I had to look at every cell in the matrix, observe that the third column in the second row is an outlier, then look at the `offsetHeight`, look back at the other numbers, recall what backtrack case 1 is and then discover what the test is doing. This did not take much time for me because I exactly knew what the code is about, so I knew what to look for, but if I did not understand the full context, then it would take much guessing to know what the test is doing).
- Are there any magic numbers that are dependent, but without any sign of such a relation?
  (In the current test, the answer is no, because changing the value of `SPACING` would change the meaning of the test. The test will likely not fail because the algorithm as a whole is correct, but the test case wouldn't be in backtrack case 1 any more).
- Is the test self-explanatory, and if not, how does it relate to the test description?
  (In the current test, because of the magic numbers and lack of comments, this point fails.)
- The test claims to be testing "backtrack case 1". Is that really true? The backtracking state is determined by the internal function logic (binary search, choice of index & helper function), and is not visible to consumers of `getVisibleElements`.

PS. Following my last bullet point, it's worth noting that `getVisibleElement`'s return value is almost trivially correct because all invisible elements are discarded in the final loop at the end of `getVisibleElements`. Even if the binary search/backtracking logic were to be completely broken, the logic at the end would save us (and hide the performance issue in the code). The correctness of the backtracking logic (and binary search) is thus important for performance, but that is not checked in the current way that the test is written. I do not expect you to write such tests, as the correctness can be checked by reviewing the code, and the extensive commentary that you added helps with that.

Reply:
I'm a little confused, sorry. Originally, you said:
> To address this maintainability issue, can you try to add tests to explicitly cover the cases that are sketched in the backtracking?

Then, you said:
> The correctness of the backtracking logic (and binary search) is thus important for performance, but that is not checked in the current way that the test is written. I do not expect you to write such tests, as the correctness can be checked by reviewing the code, and the extensive commentary that you added helps with that.

Do you or don't you want tests for each of the backtracking cases, even though the output of `getVisibleElements` can't be used to verify which path the backtracking logic took? If not, what tests were you asking me to write? (Note that the automated tests that I initially wrote do exercise all of the backtracking cases as they loop over different scroll positions, which helped me discover quite a few mistakes during development. And it's not quite true that the result of `getVisibleElements` is trivially correct even if there's a mistake in backtracking; backtracking could go back insufficiently far and thus `gVE` would return too few elements, an outcome I also observed in development.)

Reply:
Ah well, let's write those tests. Testing can be as easy as exporting `backtrackBeforeAllVisibleElements` in `ui_utils.js`, importing it in the test file and checking the result.